javascripts/tile_collision.js

```diff
@@ -0,0 +1,307 @@
+// all the tiles and their locations
+function tileInventory() {
+  var tiles = $('.tile');
+  var takenSpaces = [];
+  for (var i = 0; i < tiles.length; i++) {
+    var row = tiles[i].getAttribute("data-row");
+    var col = tiles[i].getAttribute("data-col");
+    takenSpaces.push([row, col]);
+  }
+
+  return takenSpaces;
+}
+
+// returns a fat array of all the tiles in
+// opposite order of the keystroke
+function orderTiles(keystroke) {
+  // collect all the tile locations
+  var tiles = tileInventory();
+
+  // the reaction of tiles colliding happens in the order opposite
+  // of the keystroke (e.g. up key evalautes the bottom row first)
+  var direction = function() {
+    if (keystroke === 38) { return ["r3", "r2", "r1", "r0"]; } // up
+    if (keystroke === 40) { return ["r0", "r1", "r2", "r3"]; } // down
+    if (keystroke === 37) { return ["c3", "c2", "c1", "c0"]; } // left
+    if (keystroke === 39) { return ["c0", "c1", "c2", "c3"]; } // right
+  }(); // these parens executes the function and saves the result to the variable
+
+  // determines which index to use to organize the tiles
+  // tile = `[row, column]`
+  var index = function() {
+    if (direction[0].slice(0, 1) === "r") {
+      return 0; // filters by row
+    } else if (direction[0].slice(0, 1) === "c") {
+      return 1; // filters by columns
+    }
+  }();
+
+  // select the tiles occupying the selected row/column
+  function filterByLocation(pointer) { // THREE filterByLocation(r1);
+
+    var section = tiles.filter(function(tile){ // FOUR r0,c0,r1,c1.filter for only r1,c1
+      if (tile[index] === pointer) { return tile; } // FIVE if r1 === r1, return r1, c1
+    });
+
+    return section; // SIX return r1c1
+  }
+
+  // reorganize the collected tile locations
+  // into the order of the direction array
+  var organizedTiles = function(){
+    var result = [];
+    for(var i = 0; i < direction.length; i++) { // ["r3", "r2", "r1", "r0"] } ONE
+      result = result.concat(filterByLocation(direction[i])); // same as above
+    }
+    return result;
+  };
+  return organizedTiles();
+}
+
+// move through the tiles one by one
+// figure out if they merge, move, or stay put
+function tileCollision(keystroke) {
+  var orderedTiles = orderTiles(keystroke);
+
+  var direction = function() {
```

Comment:
When you see a `switch` statement like this in multiple places with related values this is often an indication that an object wrapping that concept might be useful. For example, you could create a `Direction` object that has a `fromKeystroke` constructor. Also, using repeated string constants like `"up"`, `"down"`, etc is brittle because it makes you susceptible to bugs introduced by typos. If you use proper constants, you can let the compiler/interpreter detect invalid variable usage at least if you [`"use strict"` mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode).